Improve message when there is no spec in repos for a given dependency

manuyavuz opened the pull request:
This solves #4430.

Currently, when there is no existing specification inside any of the spec repos with given requirements, user would get the following error:

```
[!] Unable to satisfy the following requirements:
- `Lib (= 4.0.0)` required by `Podfile`
```

Now, user will get sth like this:

```
[!] Unable to find a specification for the following requirement:

	'Lib', '= 4.0.0'
```

I think this is enough for user to understand that their repo either not up-to-date or there is really no such specification with given dependency.

I actually wanted to use `Molinillo::NoSuchDependencyError` rather manually adding another message, but it's message does not contain name of the dependency, just it's version. 

I'm open to better management suggestions.

segiddins commented on lib/cocoapods/resolver.rb:
```diff
@@ -403,6 +403,10 @@ def handle_resolver_error(error)
               unless search_for(r).empty?
```
this needs to be an `if` now, since there's an `else`

segiddins commented:
hmmm, so there already should be a no such dependency error being thrown in this case? Do you think you can cook up a spec so I can have a look at what's going on?

manuyavuz commented:
Actually flow is the following when requested spec does not exist in any repo:

- `Resolver#resolve` method calls `Molinillo::Resolver#resolve` method.
- `resolve` calls `process_topmost_state`
- For this specific case, possibilities parameter is empty. 
- Because there is no `possibility`, code creates conflict using `create_conflict`

On `handle_resolve_error`, it falls into last elsif case, with `search_for` method returning empty array because there is no such spec. This is where I added the corresponding message.

I think it's better to throw a NoSuchDependencyError and improve messaging of that error, because it feels like the ideal match for that case. 
Maybe we can check if possibilities array is empty as a first step in Molinillo::Resolver#resolve method, and then throw a NoSuchDependencyError.

I modified the spec testing `raises if no such version of a dependency exists` case, and verified id new error message is actually called.

What type of spec do you think is needed?

segiddins commented:
Just a spec I can use to reproduce the issue, it sounds like a bug in molinillo to me

manuyavuz commented:
Umm ok I'll look at it
On Sun, 25 Oct 2015 at 00:32, Samuel E. Giddins <notifications@github.com>
wrote:

> Just a spec I can use to reproduce the issue, it sounds like a bug in
> molinillo to me
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CocoaPods/CocoaPods/pull/4443#issuecomment-150852749>.
>

manuyavuz commented:
I added a failing test scenario but it's not exactly the same with the original scenario.

I think original scenario include this flow:

- New Podfile and Podfile.lock files which include a version update on a dependency are pulled from project repository.
- User performs `pod install --no-repo-update`
- Because repo is not up-to-date, CP can't find specific version, but fail with an ambiguous message. Message also includes `required by 'Podfile.lock'`because Podfile.lock also include the new version.

In current scenario I'm giving an empty locked dependencies array. To generate the exact scenario, we should find a way to give a sample Dependency array.

orta commented:
```
[!] Unable to find a specification for the following requirement:

    'Lib', '= 4.0.0'
```

Is awesome, great work @manuyavuz

segiddins commented on spec/unit/resolver_spec.rb:
```diff
@@ -396,15 +396,25 @@ module Pod
         resolver = Resolver.new(config.sandbox, podfile, empty_graph, SourcesManager.all)
         e = lambda { resolver.resolve }.should.raise Informative
-        e.message.should.match(/Unable to satisfy the following requirements/)
-        e.message.should.match(/`AFNetworking \(= 3.0.1\)` required by `Podfile`/)
```
it still should say where the failing requirement came from

manuyavuz replied:
Current implementation actually satisfies the above two lines. What it should do is to satisfy the line below. I didn't add a check about the failing pod because this spec was to show the erroneous case. Correct spec will be the commented out lines.

segiddins commented:
Hmmm, why not just take the existing exception message and `gsub` it? That way we keep the metadata intact?

manuyavuz commented:
Sorry I think I don't get it. Do you mean changing the default message of VersionConflict class (`Unable to satisfy the following requirements bla bla`)?

segiddins commented:
Nope, just checking if the conflict is due to there being no possibilities, and if so raising an informative with `version_conflict.message.gsub(/unable blah blah/, 'other blah blah')`

manuyavuz commented:
Should we do this check inside Resolution#unwind_for_conflict where VersionConflict is raised, or inside Resolver#handle_resolver_error method which is my first proposal (changing message 
`if search_for(r).empty?`)?

What do you think?

segiddins commented:
No no, I think it should be handled in CocoaPods's resolver, not Molinillo

manuyavuz commented:
I've decided to give an additional information about the possible causes rather than writing the message from scratch. Error message will be like this:
```
[!] Unable to satisfy the following requirements:

- `Lib (= 4.0.0)` required by `Podfile`

None of the spec sources contain a spec satisfying the `Lib (= 4.0.0)` dependency.
Either your source repositories are not up-to-date, or there is really no such specification at all.
```

This is similar to error message structure we gave about minimum deployment target, so i think this way would be better. Any suggestions on wording?

orta commented:
Perhaps:

```
[!] Unable to satisfy the following requirements:

- `Lib (= 4.0.0)` required by `Podfile`

None of the spec sources contain a spec satisfying the `Lib (= 4.0.0)` dependency.
You have either; mistyped the name, not added the source repo that hosts the Podspec or do not have the latest versions of your source repos.
```

manuyavuz commented:
`You have either; mistyped the name or version, not added the source repo that hosts the Podspec to your Podfile, or not got the latest versions of your source repos.`

Error could be cause by mistyped version also. It's better to specify where user should add source (Podfile). Not sure for the change in the last part; my change felt more grammatically correct to me, but as an alien I may miss something.

segiddins commented:
👍🏻

manuyavuz commented:
I think it's ready to ⛵️ now!

orta commented:
awesssome

manuyavuz commented:
Thanks for help guys!
